File: backend/api/v1/serializers/admin_leads.py
from rest_framework import serializers

from billing.models import DirectSale
from accounts.models import User, UserRole
from subscriptions.models import Customer, PublicLead, Subscription


class AdminLeadListSerializer(serializers.ModelSerializer):
    product_id = serializers.IntegerField(source="product.id", read_only=True)
    product_name = serializers.CharField(source="product.name", read_only=True)
    product_code = serializers.CharField(source="product.product_code", read_only=True)
    product_base_price = serializers.DecimalField(
        source="product.base_price",
        max_digits=12,
        decimal_places=2,
        read_only=True,
    )
    assigned_to_id = serializers.IntegerField(source="assigned_to.id", read_only=True)
    assigned_to_username = serializers.CharField(
        source="assigned_to.username",
        read_only=True,
    )
    assigned_to_role = serializers.CharField(source="assigned_to.role", read_only=True)
    assigned_to_full_name = serializers.SerializerMethodField()
    converted_customer_id = serializers.IntegerField(
        source="converted_customer.id",
        read_only=True,
    )
    converted_customer_name = serializers.CharField(
        source="converted_customer.name",
        read_only=True,
    )
    converted_customer_phone = serializers.CharField(
        source="converted_customer.phone",
        read_only=True,
    )
    converted_subscription_id = serializers.IntegerField(
        source="converted_subscription.id",
        read_only=True,
    )
    converted_subscription_number = serializers.SerializerMethodField()
    converted_direct_sale_id = serializers.IntegerField(
        source="converted_direct_sale.id",
        read_only=True,
    )
    converted_direct_sale_no = serializers.SerializerMethodField()
    converted_by_id = serializers.IntegerField(source="converted_by.id", read_only=True)
    converted_by_username = serializers.CharField(
        source="converted_by.username",
        read_only=True,
    )
    converted_by_full_name = serializers.SerializerMethodField()
    party_id = serializers.SerializerMethodField()
    party_no = serializers.SerializerMethodField()
    party_display_name = serializers.SerializerMethodField()
    next_follow_up_at = serializers.SerializerMethodField()
    follow_up_state = serializers.SerializerMethodField()
    open_follow_up_count = serializers.SerializerMethodField()

    class Meta:
        model = PublicLead
        fields = (
            "id",
            "name",
            "phone",
            "city",
            "product_id",
            "product_name",
            "product_code",
            "product_base_price",
            "interested_product",
            "preferred_emi_amount",
            "status",
            "source",
            "assigned_to_id",
            "assigned_to_username",
            "assigned_to_role",
            "assigned_to_full_name",
            "assigned_at",
            "contacted_at",
            "converted_customer_id",
            "converted_customer_name",
            "converted_customer_phone",
            "converted_subscription_id",
            "converted_subscription_number",
            "converted_direct_sale_id",
            "converted_direct_sale_no",
            "converted_by_id",
            "converted_by_username",
            "converted_by_full_name",
            "party_id",
            "party_no",
            "party_display_name",
            "next_follow_up_at",
            "follow_up_state",
            "open_follow_up_count",
            "converted_at",
            "closed_at",
            "created_at",
        )

    def _crm_snapshot(self, obj):
        snapshot_map = self.context.get("lead_crm_map") or {}
        return snapshot_map.get(obj.id, {})

    def get_assigned_to_full_name(self, obj):
        user = getattr(obj, "assigned_to", None)
        if not user:
            return None

        full_name = f"{user.first_name or ''} {user.last_name or ''}".strip()
        return full_name or user.username

    def get_converted_subscription_number(self, obj):
        subscription = getattr(obj, "converted_subscription", None)
        if not subscription:
            return None
        return subscription.contract_reference or f"Subscription #{subscription.id}"

    def get_converted_direct_sale_no(self, obj):
        direct_sale = getattr(obj, "converted_direct_sale", None)
        if not direct_sale:
            return None
        return direct_sale.sale_no or f"Direct Sale #{direct_sale.id}"

    def get_converted_by_full_name(self, obj):
        user = getattr(obj, "converted_by", None)
        if not user:
            return None

        full_name = f"{user.first_name or ''} {user.last_name or ''}".strip()
        return full_name or user.username

    def get_party_id(self, obj):
        return self._crm_snapshot(obj).get("party_id")

    def get_party_no(self, obj):
        return self._crm_snapshot(obj).get("party_no")

    def get_party_display_name(self, obj):
        return self._crm_snapshot(obj).get("party_display_name")

    def get_next_follow_up_at(self, obj):
        return self._crm_snapshot(obj).get("next_follow_up_at")

    def get_follow_up_state(self, obj):
        return self._crm_snapshot(obj).get("follow_up_state", "NONE")

    def get_open_follow_up_count(self, obj):
        return self._crm_snapshot(obj).get("open_follow_up_count", 0)


class AdminLeadDetailSerializer(AdminLeadListSerializer):
    submitted_notes = serializers.CharField(source="notes", read_only=True)
    admin_notes = serializers.CharField(read_only=True)

    class Meta(AdminLeadListSerializer.Meta):
        fields = AdminLeadListSerializer.Meta.fields + (
            "submitted_notes",
            "admin_notes",
        )


class AdminLeadStatusUpdateSerializer(serializers.Serializer):
    status = serializers.ChoiceField(choices=PublicLead._meta.get_field("status").choices)


class AdminLeadAssignSerializer(serializers.Serializer):
    assigned_to = serializers.IntegerField(required=False, allow_null=True)

    def validate_assigned_to(self, value):
        if value is None:
            return None

        try:
            user = User.objects.get(pk=value)
        except User.DoesNotExist as exc:
            raise serializers.ValidationError("Selected assignee does not exist.") from exc

        if user.role not in {UserRole.ADMIN, UserRole.CASHIER, UserRole.PARTNER}:
            raise serializers.ValidationError(
                "Lead assignee must be an internal managed user."
            )

        if not user.is_active:
            raise serializers.ValidationError("Lead assignee must be active.")

        return user


class AdminLeadNoteUpdateSerializer(serializers.Serializer):
    note = serializers.CharField(allow_blank=False, trim_whitespace=True)
    mode = serializers.ChoiceField(
        choices=(
            ("append", "Append"),
            ("replace", "Replace"),
        ),
        required=False,
        default="append",
    )


class AdminLeadConversionCompleteSerializer(serializers.Serializer):
    customer_id = serializers.IntegerField(required=False, allow_null=True)
    subscription_id = serializers.IntegerField(required=False, allow_null=True)
    direct_sale_id = serializers.IntegerField(required=False, allow_null=True)

    def validate_customer_id(self, value):
        if value is None:
            return None

        try:
            return Customer.objects.get(pk=value)
        except Customer.DoesNotExist as exc:
            raise serializers.ValidationError(
                "Selected converted customer does not exist."
            ) from exc

    def validate_subscription_id(self, value):
        if value is None:
            return None

        try:
            return Subscription.objects.select_related("customer").get(pk=value)
        except Subscription.DoesNotExist as exc:
            raise serializers.ValidationError(
                "Selected converted subscription does not exist."
            ) from exc

    def validate_direct_sale_id(self, value):
        if value is None:
            return None

        try:
            return DirectSale.objects.select_related("customer").get(pk=value)
        except DirectSale.DoesNotExist as exc:
            raise serializers.ValidationError(
                "Selected direct sale does not exist."
            ) from exc

    def validate(self, attrs):
        customer = attrs.get("customer_id")
        subscription = attrs.get("subscription_id")
        direct_sale = attrs.get("direct_sale_id")

        if customer is None and subscription is None and direct_sale is None:
            raise serializers.ValidationError(
                "Select the created customer, subscription, or direct sale before completing conversion."
            )

        if customer is not None and subscription is not None:
            if subscription.customer_id != customer.id:
                raise serializers.ValidationError(
                    "Selected subscription does not belong to the selected customer."
                )

        if customer is not None and direct_sale is not None and direct_sale.customer_id:
            if direct_sale.customer_id != customer.id:
                raise serializers.ValidationError(
                    "Selected direct sale does not belong to the selected customer."
                )

        if subscription is not None and direct_sale is not None:
            if direct_sale.customer_id and direct_sale.customer_id != subscription.customer_id:
                raise serializers.ValidationError(
                    "Selected direct sale does not belong to the selected subscription customer."
                )

        return attrs

File: frontend/src/services/admin-leads.ts
import { apiFetch } from "@/lib/api";

export type AdminLeadStatus =
  | "NEW"
  | "IN_PROGRESS"
  | "CONTACTED"
  | "CONVERTED"
  | "CLOSED";

export type AdminLeadIntent =
  | "GENERAL"
  | "QUOTATION"
  | "ESTIMATE"
  | "DIRECT_SALE"
  | "SUBSCRIPTION";

export type AdminLeadRow = {
  id: number;
  name: string;
  phone: string;
  city?: string;
  product_id?: number | null;
  product_name?: string | null;
  product_code?: string | null;
  product_base_price?: string | null;
  interested_product?: string;
  preferred_emi_amount?: string | null;
  status: AdminLeadStatus;
  intent?: AdminLeadIntent;
  source?: string;
  follow_up_required?: boolean;
  follow_up_on?: string | null;
  follow_up_note?: string;
  assigned_to_id?: number | null;
  assigned_to_username?: string | null;
  assigned_to_role?: string | null;
  assigned_to_full_name?: string | null;
  assigned_at?: string | null;
  contacted_at?: string | null;
  converted_customer_id?: number | null;
  converted_customer_name?: string | null;
  converted_customer_phone?: string | null;
  converted_subscription_id?: number | null;
  converted_subscription_number?: string | null;
  converted_direct_sale_id?: number | null;
  converted_direct_sale_no?: string | null;
  converted_by_id?: number | null;
  converted_by_username?: string | null;
  converted_by_full_name?: string | null;
  party_id?: number | null;
  party_no?: string | null;
  party_display_name?: string | null;
  next_follow_up_at?: string | null;
  follow_up_state?: "NONE" | "DUE" | "SCHEDULED";
  open_follow_up_count?: number;
  converted_at?: string | null;
  closed_at?: string | null;
  created_at?: string | null;
};

export type AdminLeadDetail = AdminLeadRow & {
  submitted_notes?: string;
  admin_notes?: string;
};

export type AdminLeadSummary = {
  total: number;
  new: number;
  in_progress: number;
  contacted: number;
  converted: number;
  closed: number;
  assigned: number;
  unassigned: number;
};

export type AdminLeadListResponse = {
  count: number;
  summary: AdminLeadSummary;
  results: AdminLeadRow[];
};

export type AdminLeadListQuery = {
  q?: string;
  status?: AdminLeadStatus | "";
  intent?: AdminLeadIntent | "";
  assignee?: string;
  date_from?: string;
  date_to?: string;
};

function buildQuery(params: AdminLeadListQuery): string {
  const search = new URLSearchParams();

  if (params.q?.trim()) search.set("q", params.q.trim());
  if (params.status) search.set("status", params.status);
  if (params.intent) search.set("intent", params.intent);
  if (params.assignee?.trim()) search.set("assignee", params.assignee.trim());
  if (params.date_from?.trim()) search.set("date_from", params.date_from.trim());
  if (params.date_to?.trim()) search.set("date_to", params.date_to.trim());

  const query = search.toString();
  return query ? `?${query}` : "";
}

export async function listAdminLeads(
  query: AdminLeadListQuery = {}
): Promise<AdminLeadListResponse> {
  return apiFetch<AdminLeadListResponse>(`/admin/leads/${buildQuery(query)}`);
}

export async function createAdminLead(payload: {
  name: string;
  phone: string;
  email?: string;
  city?: string;
  product_id?: number | null;
  interested_product?: string;
  preferred_emi_amount?: string;
  notes?: string;
  admin_notes?: string;
  source?: string;
  intent?: AdminLeadIntent;
  follow_up_required?: boolean;
  follow_up_on?: string | null;
  follow_up_note?: string;
  assigned_to?: number | null;
}): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>("/admin/leads/", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getAdminLead(id: number | string): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>(`/admin/leads/${id}/`);
}

export async function updateAdminLeadStatus(
  id: number | string,
  status: AdminLeadStatus
): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>(`/admin/leads/${id}/status/`, {
    method: "POST",
    body: JSON.stringify({ status }),
  });
}

export async function updateAdminLeadAssignee(
  id: number | string,
  assigned_to: number | null
): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>(`/admin/leads/${id}/assign/`, {
    method: "POST",
    body: JSON.stringify({ assigned_to }),
  });
}

export async function updateAdminLeadNotes(
  id: number | string,
  payload: {
    note: string;
    mode: "append" | "replace";
  }
): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>(`/admin/leads/${id}/notes/`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function completeAdminLeadConversion(
  id: number | string,
  payload: {
    customer_id?: number | null;
    subscription_id?: number | null;
    direct_sale_id?: number | null;
  }
): Promise<AdminLeadDetail> {
  return apiFetch<AdminLeadDetail>(`/admin/leads/${id}/convert/`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
